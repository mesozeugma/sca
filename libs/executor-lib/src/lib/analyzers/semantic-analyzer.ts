import klaw from 'klaw';
import path from 'node:path';
import { z } from 'zod';

import { AnalysisOutput } from '../analysis/analysis-output';
import { OpensearchIndex } from '../opensearch/opensearch-index';
import { Utils } from '../utils';

import { Analyzer } from './analyzer';

const SemanticSymbol = z.object({}).passthrough();
const SemanticJSONSymbols = z
  .object({
    files: z.array(
      z
        .object({
          symbols: z.array(SemanticSymbol).default([]),
        })
        .strip()
    ),
  })
  .strip();

const SemanticResult = z
  .object({
    file: z.string(),
    symbol: SemanticSymbol,
  })
  .strict();
type SemanticResult = z.infer<typeof SemanticResult>;

export class SemanticAnalyzer implements Analyzer {
  private filterPath(file: string) {
    const fileExtension = path.extname(file);
    return ['.java', '.py'].some((validExtension) =>
      fileExtension.endsWith(validExtension)
    );
  }

  private async seamntic(repositoryPath: string): Promise<SemanticResult[]> {
    if (!repositoryPath) {
      throw Error('Repository path not set');
    }
    const result: SemanticResult[] = [];
    for await (const file of klaw(repositoryPath, {
      depthLimit: -1,
    })) {
      if (!this.filterPath(file.path)) continue;

      const { stdout } = await Utils.executeCommand('semantic', [
        'parse',
        '--json-symbols',
        file.path,
      ]);
      const jsonSymbols = await SemanticJSONSymbols.parseAsync(
        JSON.parse(stdout)
      );
      if (jsonSymbols.files.length === 0) continue;
      jsonSymbols.files[0].symbols.forEach((element) => {
        result.push({
          file: path.relative(repositoryPath, file.path),
          symbol: element,
        });
      });
    }

    return result;
  }

  async readRepo(repositoryPath: string) {
    const items = await this.seamntic(repositoryPath);
    return items.map((e) =>
      AnalysisOutput.create(OpensearchIndex.ANALYSIS_RESULT_SEMANTIC, e)
    );
  }
}
