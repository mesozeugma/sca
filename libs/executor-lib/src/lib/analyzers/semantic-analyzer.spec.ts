import { SemanticAnalyzer } from './semantic-analyzer';

test('SemanticAnalyzer', async () => {
  const analyzer = new SemanticAnalyzer();
  const result = await analyzer.readRepo(
    './libs/executor-lib/code-samples/java/'
  );
  expect(result).toMatchSnapshot();
});
