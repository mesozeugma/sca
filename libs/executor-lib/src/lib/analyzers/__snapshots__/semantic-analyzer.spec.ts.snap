// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SemanticAnalyzer 1`] = `
[
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "symbol": {
        "kind": "Class",
        "line": "public class HelloWorld extends Thread implements Runnable {",
        "nodeType": "DEFINITION",
        "span": {
          "end": {
            "column": 24,
            "line": 6,
          },
          "start": {
            "column": 14,
            "line": 6,
          },
        },
        "symbol": "HelloWorld",
        "syntaxType": "CLASS",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 23,
            "line": 5,
          },
          "start": {
            "column": 13,
            "line": 5,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "symbol": {
        "kind": "Method",
        "line": "public static void main(String[] args) {",
        "nodeType": "DEFINITION",
        "span": {
          "end": {
            "column": 28,
            "line": 7,
          },
          "start": {
            "column": 24,
            "line": 7,
          },
        },
        "symbol": "main",
        "syntaxType": "METHOD",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 27,
            "line": 6,
          },
          "start": {
            "column": 23,
            "line": 6,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "symbol": {
        "kind": "Call",
        "line": "run();",
        "nodeType": "REFERENCE",
        "span": {
          "end": {
            "column": 12,
            "line": 8,
          },
          "start": {
            "column": 9,
            "line": 8,
          },
        },
        "symbol": "run",
        "syntaxType": "CALL",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 11,
            "line": 7,
          },
          "start": {
            "column": 8,
            "line": 7,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "symbol": {
        "kind": "Method",
        "line": "public void run() {",
        "nodeType": "DEFINITION",
        "span": {
          "end": {
            "column": 20,
            "line": 11,
          },
          "start": {
            "column": 17,
            "line": 11,
          },
        },
        "symbol": "run",
        "syntaxType": "METHOD",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 19,
            "line": 10,
          },
          "start": {
            "column": 16,
            "line": 10,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "symbol": {
        "kind": "Call",
        "line": "System.out.println("Hello World");",
        "nodeType": "REFERENCE",
        "span": {
          "end": {
            "column": 27,
            "line": 13,
          },
          "start": {
            "column": 20,
            "line": 13,
          },
        },
        "symbol": "println",
        "syntaxType": "CALL",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 26,
            "line": 12,
          },
          "start": {
            "column": 19,
            "line": 12,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "HelloWorld.java",
      "symbol": {
        "kind": "Implementation",
        "line": "public class HelloWorld extends Thread implements Runnable {",
        "nodeType": "REFERENCE",
        "span": {
          "end": {
            "column": 59,
            "line": 6,
          },
          "start": {
            "column": 51,
            "line": 6,
          },
        },
        "symbol": "Runnable",
        "syntaxType": "IMPLEMENTATION",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 58,
            "line": 5,
          },
          "start": {
            "column": 50,
            "line": 5,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "nested/IsEven.java",
      "symbol": {
        "kind": "Class",
        "line": "public class IsEven {",
        "nodeType": "DEFINITION",
        "span": {
          "end": {
            "column": 20,
            "line": 5,
          },
          "start": {
            "column": 14,
            "line": 5,
          },
        },
        "symbol": "IsEven",
        "syntaxType": "CLASS",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 19,
            "line": 4,
          },
          "start": {
            "column": 13,
            "line": 4,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
  AnalysisOutput {
    "data": {
      "file": "nested/IsEven.java",
      "symbol": {
        "kind": "Method",
        "line": "public static boolean  isEven(int number) {",
        "nodeType": "DEFINITION",
        "span": {
          "end": {
            "column": 34,
            "line": 6,
          },
          "start": {
            "column": 28,
            "line": 6,
          },
        },
        "symbol": "isEven",
        "syntaxType": "METHOD",
        "utf16CodeUnitSpan": {
          "end": {
            "column": 33,
            "line": 5,
          },
          "start": {
            "column": 27,
            "line": 5,
          },
        },
      },
    },
    "index": "public_sca_analysis_result_semantic",
  },
]
`;
